tracer/src/Datadog.Tracer.Native/clr_helpers.cpp

```diff
@@ -1018,7 +1018,318 @@ HRESULT HasAsyncStateMachineAttribute(const ComPtr<IMetaDataImport2>& metadataIm
+                    {
+                        const auto& assemblyRefMetadata =
+                            GetReferencedAssemblyMetadata(candidateAssemblyImport, exportedTypeContainerToken);
+                        if (assemblyRefMetadata.name == shared::EmptyWStr)
+                        {
+                            Logger::Warn(
+                                "[ResolveTypeInternal] GetResolutionScopeNameForAssemblyRefTok has failed with: ", shared::WSTRING(refTypeName.data()));
+                            continue;
+                        }
+
+                        hr = ResolveTypeInternal(info, loadedModules, refTypeName, mdExportedTypeNil, assemblyRefMetadata.name,
+                                                resolvedTypeDefToken, resolvedTypeDefMetadataImport);
+                        if (FAILED(hr))
+                        {
+                            Logger::Warn(
+                                "[ResolveTypeInternal] Recursive call to ResolveTypeInternal has failed with: ", shared::WSTRING(refTypeName.data()));
+                            continue;
+                        }
+
+                        return hr;
+                    }
+                }
+            }
+        }
+    }
+
+    if (isModuleFound)
+    {
+        return S_OK;
+    }
+    
+    resolvedTypeDefToken = mdTokenNil;
+    Logger::Error("[ResolveTypeInternal] ResolveTypeInternal has failed. Reason: Module not found in the loaded modules for type name: ", shared::WSTRING(refTypeName.data()));
+    return E_NOTIMPL;
+}
+
+/**
+ Resolves TypeRef to TypeDef.
+ Note: If the module where the TypeRef is defined is not loaded, E_FAIL will return.
+ */
+HRESULT ResolveType(ICorProfilerInfo4* info,
+                    const ComPtr<IMetaDataImport2>& metadata_import,
+                    const ComPtr<IMetaDataAssemblyImport>& assembly_import,
+                    const mdTypeRef typeRefToken,
+                    mdTypeDef& resolvedTypeDefToken,
+                    ComPtr<IMetaDataImport2>& resolvedMetadataImport)
+{
+    mdToken resolutionScope = mdTokenNil; // will hold either AssemblyRef or ModuleRef token
+    mdToken enclosingType = mdTokenNil;
+    ULONG nameSize = 0;
+
+    auto hr = metadata_import->GetTypeRefProps(typeRefToken, &resolutionScope, nullptr, 0, &nameSize);
+    if (FAILED(hr) || resolutionScope == mdTokenNil)
+    {
+        Logger::Error("[ResolveType] GetTypeRefProps [1] has failed. typeRefToken: ", typeRefToken);
+        return E_FAIL;
+    }
+    // get the type name & resolution scope
+    std::vector<WCHAR> refTypeName(nameSize, 0);
+    hr = metadata_import->GetTypeRefProps(typeRefToken, &resolutionScope, refTypeName.data(),
+                                                          kNameMaxSize, &nameSize);
+    if (FAILED(hr) || resolutionScope == mdTokenNil)
+    {
+        Logger::Error("[ResolveType] GetTypeRefProps [2] has failed. typeRefToken: ", typeRefToken);
+    }
+    IfFailRet(hr);
```

Comment:
Can you try to retrieve the `refTypeName` in the first call with a default buffer size? and if fails try with the actual buffer size returned by the first call? Depending of the buffer size we may skip the second call at all.

Reply:
I tend to agree with you, it was a bad habit from my past experience with embedded software 😅 .
Deduplicated in ef502336ac5073c545e222bc6396635a9d76288f. It's totally fine calling it only once, it's also what we are doing elsewhere in the codebase.
Thanks!